GitUp/Application/WelcomeWindow.m

```diff
@@ -97,49 +97,144 @@ - (void)concludeDragOperation:(id<NSDraggingInfo>)sender {
+@implementation WelcomeWindow
+
+#pragma mark - Setup
+- (void)setup {
+  self.opaque = NO;
+  self.backgroundColor = [NSColor clearColor];
+  self.movableByWindowBackground = YES;
+}
+
+- (void)awakeFromNib {
+  [self setup];
+}
+
+#pragma mark - Window
+- (BOOL)canBecomeKeyWindow {
+  return YES;
+}
+
+@end
+
+typedef NS_ENUM(NSInteger, WelcomeWindowControllerWindowState) {
+  WelcomeWindowControllerWindowStateNotActivated,
+  WelcomeWindowControllerWindowStateClosed,
+  WelcomeWindowControllerWindowStateShouldBeOpened
+};
+@interface WelcomeWindowControllerModel ()
+@property (assign, nonatomic, readwrite) WelcomeWindowControllerWindowState state;
+@end
+
+@implementation WelcomeWindowControllerModel
+#pragma mark - Setters
+- (void)setShouldShow {
+  self.state = WelcomeWindowControllerWindowStateShouldBeOpened;
+}
+- (void)setShouldHide {
+  self.state = WelcomeWindowControllerWindowStateClosed;
+}
+
+#pragma mark - Getters
+- (BOOL)notActivedYet {
+  return self.state == WelcomeWindowControllerWindowStateNotActivated;
+}
+- (BOOL)shouldShow {
+  return self.state == WelcomeWindowControllerWindowStateShouldBeOpened;
+}
+@end
+
+@interface WelcomeWindowController ()
+@property(nonatomic, weak) IBOutlet NSButton *closeButton;
 @property(nonatomic, weak) IBOutlet NSPopUpButton* recentPopUpButton;
 @property(nonatomic, weak) IBOutlet GILinkButton* twitterButton;
 @property(nonatomic, weak) IBOutlet GILinkButton* forumsButton;
 
-@property (weak, nonatomic, readonly) AppDelegate *appDelegate;
-@property (weak, nonatomic, readwrite) IBOutlet WelcomeWindowView *destinationView;
+@property (strong, nonatomic, readwrite) WelcomeWindowControllerModel *model;
 @end
 
-@implementation WelcomeWindow
+@implementation WelcomeWindowController
 
-#pragma mark - Accessors
-- (AppDelegate *)appDelegate {
-  return [AppDelegate sharedDelegate];
+#pragma mark - Initialization
+- (instancetype)init {
+  return [[super initWithWindowNibName:@"Welcome"] configuredWithModel:[[WelcomeWindowControllerModel alloc] init]];
 }
 
-#pragma mark - Actions
+#pragma mark - Setup
+- (void)setupUIElements {
+  self.twitterButton.textAlignment = NSLeftTextAlignment;
+  self.twitterButton.textFont = [NSFont boldSystemFontOfSize:11];
+  self.forumsButton.textAlignment = NSLeftTextAlignment;
+  self.forumsButton.textFont = [NSFont boldSystemFontOfSize:11];
+  
+  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(willShowPopUpMenu) name:NSPopUpButtonWillPopUpNotification object:self.recentPopUpButton];
+  
+  self.closeButton.action = @selector(closeButtonPressed);
+  self.twitterButton.action = @selector(openTwitter);
+  
+  self.closeButton.target = self;
+  self.twitterButton.target = self;
+}
+
+#pragma mark - Window Lifecycle
+- (void)windowDidLoad {
+  [super windowDidLoad];
+  [self setupUIElements];
+}
+
+#pragma mark - Configuration
+- (instancetype)configuredWithModel:(WelcomeWindowControllerModel *)model {
+  self.model = model;
+  return self;
+}
```

Comment:
Just do standard initialization in the init method, don't make it a pseudo init returning instancetype:
```
- (instancetype)init {
  if (self = [self initWithWindowNibName:@"Welcome"]) {
    _model = [[WelcomeWindowControllerModel alloc] init]];
  }
  return self;
 }
```
I'm also not really convinced this model object adds anything other than complexity (it's never going to be reused anywhere else).

Reply:
@lucasderraugh Fixed.